### ISSUE TITLE
As a SO, I want to be able to do basic A/B testing

### ISSUE THREAD

**Issue by kreynen** [CuBoulder/express#1015]
## Context

Carolyn asked for this specifically for YouTube videos where they would monitor the drop off rate based on different edits.  That usecase is easy enough to do with a few lines of javascript that should all the even second page requests video A and odd second loads video B.

If we did this at the node/path level we'd be able to facilitate some basic A/B testing with just about anything Express supported... layouts, designs, menu options, etc. 

We need a solution that works at the path level.  This will likely need to work with #921 or https://www.google.com/analytics/tag-manager/

## Process

NEEDS TO BE DEFINED BASED ON APPROACH

## Expected result

Visitors to a URL like colorado.edu/greatest-ad-ever would see either the actual content configured for that path or the B version of that.  The percentage of users seeing each option needs to be configurable.  

## Current result

NA

## Possible Fix

https://www.drupal.org/project/abjs -> maintained and D8 release, but it is this developers only project on Drupal.org
https://www.drupal.org/project/abtest -> no maintained, but originally written by dsnopek 
https://www.drupal.org/sandbox/sandykadam/2189611

https://www.drupal.org/project/personalize -> Most comprehensive toolset I've seen. Breaks down personalization into logical components.

**Comment by alexfinnarn** [comment]
One important consideration is to think about Varnish. Right now, Varnish caches a page for five minutes. If we only do something on the server-side, then users have to wait five minutes before getting the next version of the page. If we up the Varnish TTL, the problem would only be worse. 

Because of this, caching is a huge concern. Implementing the changes via JS mitigates this issue.

**Comment by alexfinnarn** [comment]
Review of AB JS module...

- You end up adding "experiences" which are the parts of the page that would change. The experiences are added via a text field and would require exact knowledge of page structure to work. 
- Every test needs a "condition" to act upon. Even if you want this test to run on all pages, you have to make a condition for that. If you only want it to run on a certain page, you have to include that via JS. Creating dynamic logic for adding to multiple pages can get complex pretty fast. 
- An end user can't create their own tests since you have to allow the user to write JS. Users could activate different tests, but the module would have to be hacked or separate forms created to allow the user to even pick fractions for A/B distribution.
- No API hooks. A completely self-contained module that uses no Drupalisms.

**Comment by alexfinnarn** [comment]
Review of Personalization module...

- Has an admin mode akin to GA where admins aren't tracked in A/B testing.
- Uses ctools plugins for all of its component declarations meaning it is easy to add custom plugins according to our needs. 
- Has a well-defined architecture of Decision Agents, Option Sets, Executors, Goals, and Visitor Contexts.
- Can schedule when to start and stop experiments. 
- Would need to create a visitor context for displaying content via fractions. 
- The most comprehensive A/B personalization module I've seen in D7.

**Comment by alexfinnarn** [comment]
After trying to extend the personalize suite of modules, I decided to roll my own. I would have liked to use those modules since they already include several different ways of performing tests on your site. 

With the idea of making this a contrib module for D8 (and any D7 module we make now should consider this), it was easier to start with the video team's request. I hope to make this into a D8 module once ironed out. 

The current solution allows a user to create an "A/B Block" where they can add unlimited field collections of "block options" which are beans with a percentage attached. The percentage is used to calculate which option should be shown on page load. 

I did plan to report something back in the form of a counter so a user could see if the percentages matched what the users were actually being shown. You could report other things back at that point too and end up with a progressive A/B test towards some final winning block option. 

I think adjusting the percentage during a test experiment is a common thing in A/B tests, but those considerations would be more of a Phase II task. 

The module could also be abstracted to have a base module and then a bean/block sub-module that adds the block option set for comparisons. Similarly, the percentage way of showing the blocks could be abstracted to a reaction option...kind of like Rules condition/reaction dichotomy.

**Comment by alexfinnarn** [comment]
![screen shot 2017-04-18 at 3 41 09 pm](https://cloud.githubusercontent.com/assets/3640707/25154332/7b50cdf2-244d-11e7-8bc1-1770beefaf54.png)

**Comment by kreynen** [comment]
This code now part of https://github.com/CuBoulder/cu_digital_campaign_bundle, but won't be implemented until after 2.4.0

**Comment by jwfuller** [comment]
Can't let the cache lifetimes get too short

**Comment by alexfinnarn** [comment]
I'm calling this issue a success now that it is out in the wild and there are no updated comments or follow-ups.
